### PR TITLE
Feature/grpc bgp demo

### PR DESCRIPTION
**Pull request by RichWellum**
Minitest happy:

54 runs, 141 assertions, 0 failures, 0 errors, 18 skips
Coverage report generated for MiniTest to /ws/rwellum-rtp/github/cisco-network-node-utils/tests/coverage. 0.0 / 0.0 LOC (100.0%) covered.

Rubocop clean on modified files.

-_\- mode: compilation; default-directory: "/ws/rwellum-rtp/github/cisco-network-node-utils/lib/cisco_node_utils/" -_-
Compilation started at Tue Nov  3 15:25:11

rubocop --format emacs /ws/rwellum-rtp/github/cisco-network-node-utils/

Compilation finished at Tue Nov  3 15:25:39

### CODE REVIEW

File: lib/cisco_node_utils/bgp.rb [hunk 1]
```diff
@@ -84,7 +84,9 @@ def self.enabled
     end
 
     def self.enable(state='')
-      config_set('bgp', 'feature', state: state)
+      # IOS XR has no '[no] feature xyz'.
+      config_set('bgp', 'feature', state: state) if
+        node.client.api == 'NXAPI' || state == 'no'
```
Comment (glennmatthews) [inline lib/cisco_node_utils/bgp.rb:89]: seems to me like the `state == 'no'` check is incorrect - you should _only_ try to invoke the 'feature bgp' command on NXAPI, regardless of state

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 2]
```diff
@@ -1,109 +1,197 @@
...
+    config_set_append:
+      - " vrf <vrf>"
+  nxapi:
+    config_get: "show running bgp all"
+    config_get_token: '/^router bgp <asnum>$/'
+    config_get_token_append:
+      - '/^vrf <vrf>$/'
+    config_set: "router bgp <asnum>"
+    config_set_append:
+      - "vrf <vrf>"
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:17]: Values that are the same for both can be shared between grpc and nxapi:

``` yaml
_template:
  config_get_token: '/^router bgp <asnum>$/'
  config_get_token_append:
    - '/^vrf <vrf>$/'
  config_set: "router bgp <asnum>"
  grpc:
    config_get: "show running-config router bgp"
    # Note the space before 'vrf' - seems like grpc
    # is currently indentation sensitive.
    config_set_append:
      - " vrf <vrf>"
  nxapi:
    config_get: "show running bgp all"
    config_set_append:
      - "vrf <vrf>"
```

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 3]
```diff
@@ -1,109 +1,197 @@
...
-  config_set_append: '<state> bestpath always-compare-med'
-  default_value: false
+  grpc:
+    config_get_token_append: '/^bgp bestpath med always$/'
+    config_set_append: ' <state> bgp bestpath med always'
+    default_value: false
+  nxapi:
+    config_get_token_append: '/^bestpath always-compare-med$/'
+    config_set_append: '<state> bestpath always-compare-med'
+    default_value: false
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: ditto - `default_value` is `false` for both so it doesn't need to be repeated. Same suggestion throughout below.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 4]
```diff
@@ -1,109 +1,197 @@
...
+  grpc:
+    config_get_token_append: '/^bgp confederation identifier (\d+.?\d+?)$/'
+    config_set_append: ' <state> bgp confederation identifier <id>'
+    default_value: ""
+  nxapi:
+    config_get_token_append: '/^confederation identifier (\d+.?\d+?)$/'
+    config_set_append: '<state> confederation identifier <id>'
+    default_value: ""
 
+#TODO regexp is broken
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: can you revert this part or comment out the broken part?

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 5]
```diff
@@ -1,109 +1,197 @@
...
+    config_set_append: '<state> enforce-first-as'
+    default_value: true
 
 feature:
-  config_get: "show running bgp"
-  config_get_token: '/^feature bgp$/'
-  config_set: "<state> feature bgp"
+  grpc:
+    config_get: "show running-config router bgp"
+    config_get_token: '/^router bgp$/'
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:123]: I thought we'd discussed that the right thing to do for 'feature' was to make it a no-op on XR, rather than making it a clone of 'router'?

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 6]
```diff
@@ -1,109 +1,197 @@
...
+  grpc:
+    config_get_token_append: '/^bgp graceful-restart$/'
+    config_set_append: ' <state> bgp graceful-restart'
+    default_value: true
+  nxapi:
+    config_get_token_append: '/^graceful-restart$/'
+    config_set_append: '<state> graceful-restart'
+    default_value: true
 
+#Does not exist in IOS XR
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Can you mark this as a 'TODO:'? This is one that will need to be updated once we have proper `exclude_platform` capability in the YAML.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 4 above)
Comment (mikewiebe) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Yeah.. I think the regexp should be this instead.

(\d+|\d+.\d+)

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 7]
```diff
@@ -111,30 +199,57 @@ neighbor_fib_down_accelerate:
...
-  config_get_token: '/^router bgp (\d+)$/'
-  config_set: "<state> router bgp <asnum>"
+  grpc:
+    config_get: "show running router bgp"
+    config_get_token: '/^router bgp (\d+)$/'
+    config_set: "<state> router bgp <asnum>"
+  nxapi:
+    config_get: "show running bgp"
+    config_get_token: '/^router bgp (\d+)$/'
+    config_set: "<state> router bgp <asnum>"
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: shared information:

```
router:
  config_get_token: '/^router bgp (\d+)$/'
  config_set: "<state> router bgp <asnum>"
  grpc:
    config_get: "show running router bgp"
  nxapi:
    config_get: "show running bgp"
```

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 8]
```diff
@@ -111,30 +199,57 @@ neighbor_fib_down_accelerate:
...
-  default_value: false
+  grpc:
+    config_get: "show running router bgp"
+    config_get_token: ['/^router bgp %s$/i', '/^shutdown$/']
+    config_set: ["router bgp <asnum>", "<state> shutdown"]
+    default_value: false
+  nxapi:
+    config_get: "show running bgp"
+    config_get_token: ['/^router bgp %s$/i', '/^shutdown$/']
+    config_set: ["router bgp <asnum>", "<state> shutdown"]
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:215]: shared values yadda yadda

File: lib/cisco_node_utils/cmd_ref/bgp.yaml [hunk 9]
```diff
@@ -153,9 +268,14 @@ timer_bgp_keepalive:
   default_value: 60
 
 timer_bgp_keepalive_hold:
-  config_get_token_append: '/^timers bgp (\d+) (\d+)$/'
-  config_set_append: '<state> timers bgp <keepalive> <hold>'
+  #Looks the same
```
Comment (glennmatthews) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: please fix

File: tests/test_router_bgp.rb [hunk 10]
```diff
@@ -20,17 +20,38 @@
 require_relative 'ciscotest'
 require_relative '../lib/cisco_node_utils/bgp'
 
+XR_NOT_SUPPORTED     = 'Not supported in IOS XR'
+XR_NO_VRF_SUPPORT    = 'Feature not supported in a vrf in IOS XR'
+XR_SUPPORTED_BROKEN  = 'Supported in IOS XR - needs further work'
+
+def test_add_routerid_rdauto(asnum, vrf)
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: note that any function starting with `test_` is assumed to be a test case by Minitest - if this is just a helper method for other test cases, please change its name.

File: tests/test_router_bgp.rb [hunk 11]
```diff
@@ -130,9 +154,16 @@ def test_routerbgp_create_valid_no_feature
     refute_nil(line, "Error: 'router bgp #{asnum}' not configured")
     bgp.destroy
 
-    s = @device.cmd('show run all | no-more')
-    line = /"feature bgp"/.match(s)
-    assert_nil(line, "Error: 'feature bgp' still configured")
+    if node.client.api == 'gRPC'
+      config('term len 0')
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: basetest.rb already does this when it connects, so it shouldn't be needed here.

File: tests/test_router_bgp.rb [hunk 12]
```diff
@@ -192,15 +224,19 @@ def test_routerbgp_set_get_bestpath
       assert(bgp.bestpath_cost_community_ignore, "vrf #{vrf}: "\
              'bgp bestpath_cost_community_ignore should be enabled')
       bgp.bestpath_med_confed = true
-      assert(bgp.bestpath_med_confed, "vrf #{vrf}: "\
-             'bgp bestpath_med_confed should be enabled')
+      if node.client.api == 'NXAPI' ||
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb:226]: These checks are added because only some of the bestpath options are working yet in the node utils for XR, not because of any inherent NX-exclusivity of these options, right? If so, can you please mark these with 'TODO:'?

File: tests/test_router_bgp.rb [hunk 13]
```diff
@@ -282,9 +325,11 @@ def test_routerbgp_set_get_cluster_id
         vrf = 'default'
         bgp = RouterBgp.new(asnum)
       else
+        skip(XR_NO_VRF_SUPPORT) if node.client.api == 'gRPC'
         asnum = 99
         vrf = 'yamllll'
         bgp = RouterBgp.new(asnum, vrf)
+        test_add_routerid_rdauto(asnum, vrf)
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: shouldn't be needed since you're skipping this test case on XR, right?

Same comment throughout this file.

File: tests/test_router_bgp.rb [hunk 14]
```diff
@@ -282,9 +325,11 @@ def test_routerbgp_set_get_cluster_id
         vrf = 'default'
         bgp = RouterBgp.new(asnum)
       else
+        skip(XR_NO_VRF_SUPPORT) if node.client.api == 'gRPC'
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: Note that this will cause the entire test to report as skipped (including the default VRF part). If this isn't supported on XR as an OS limitation, the better thing to do would be either to refactor the test into two test cases (and skip only the non-default-VRF test case) or simply do `next` here instead of `skip(reason)` to bypass this section of the loop only.

File: tests/test_router_bgp.rb [hunk 15]
```diff
@@ -342,9 +388,11 @@ def test_routerbgp_set_get_graceful_restart
...
         asnum = 99
         vrf = 'yamllll'
         bgp = RouterBgp.new(asnum, vrf)
+        test_add_routerid_rdauto(asnum, vrf)
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: ditto unnecessary since skipping already on XR

File: tests/test_router_bgp.rb [hunk 16]
```diff
@@ -389,7 +440,9 @@ def test_routerbgp_default_graceful_restart
     assert_equal(300, bgp.default_graceful_restart_timers_stalepath_time,
                  "bgp graceful restart default timer value should be '300'")
     refute(bgp.default_graceful_restart_helper,
-           'graceful restart helper default value should be enabled = false')
+           'graceful restart helper default value ' \
+           'should be enabled = false') if
+      node.client.api == 'NXAPI'
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb]: Ick, that's hard to parse. Should this be marked as a TODO or is this something that just plain doesn't work on XR?

```
if node.client.api == 'NXAPI'
  refute(bgp.default_graceful_restart_helper, ....
end
```

File: tests/test_router_bgp.rb [hunk 17]
```diff
@@ -20,17 +20,38 @@
 require_relative 'ciscotest'
 require_relative '../lib/cisco_node_utils/bgp'
 
+XR_NOT_SUPPORTED     = 'Not supported in IOS XR'
```
Comment (glennmatthews) [inline tests/test_router_bgp.rb:23]: Are there really so many commands that work in NXOS but not in IOS XR? I though XR was supposed to be the gold standard for BGP.

File: lib/cisco_node_utils/bgp.rb (hunk 1 above)
Comment (RichWellum) [inline lib/cisco_node_utils/bgp.rb:89]: This hits the 'feature' block of yaml amd in the XR no case only, I do a 'no router bgp'. So I am using it in a way. Otherwise changing the ASN testcases don't work well.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 2 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:17]: I'll clean this up for the whole file.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 3 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Ack.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 4 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: It's referring to confederation peers below. Will put a better comment in.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 5 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:123]: I'm not sure - it seemed reasonable to equate 'no feature bgp' with 'no router bgp' to me in the end.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 6 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Ack.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 7 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Ack.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 8 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml:215]: Ack.

File: lib/cisco_node_utils/cmd_ref/bgp.yaml (hunk 9 above)
Comment (RichWellum) [inline lib/cisco_node_utils/cmd_ref/bgp.yaml]: Ack.

File: tests/test_router_bgp.rb (hunk 17 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb:23]: I think the skip count is high because of what you pointed out before - I am skipping whole test cases. But I'll do an assessment and get back to you once I've changed those tcs.

File: tests/test_router_bgp.rb (hunk 10 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Ack.

File: tests/test_router_bgp.rb (hunk 11 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Ack.

File: tests/test_router_bgp.rb (hunk 12 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb:226]: No in this case and a few others they worked in IOS XR at the global level (vrf = default) but not under a vrf. I'll add an appropriate comment to these.

File: tests/test_router_bgp.rb (hunk 14 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Ack.

File: tests/test_router_bgp.rb (hunk 13 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Ack. Applied to whole file.

File: tests/test_router_bgp.rb (hunk 15 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Ack.

File: tests/test_router_bgp.rb (hunk 16 above)
Comment (RichWellum) [inline tests/test_router_bgp.rb]: Plain doesn't work on XR. Reason I did it like the above is because rubocop complained about usign a guard clause.

File: lib/cisco_node_utils/bgp.rb (hunk 1 above)
Comment (glennmatthews) [inline lib/cisco_node_utils/bgp.rb:89]: OK - not blocking but maybe change the above comment:

```
# IOS XR has no '[no] feature bgp' command so:
#   enable == false: do 'no router bgp' instead
#   enable == true:  no-op
```

**Comment by glennmatthews** [comment]
:+1:
